sdk/core/src/context.rs

```diff
@@ -19,9 +32,23 @@ impl Context {
     pub fn new() -> Self {
         Self {
             type_map: HashMap::new(),
+            previous_context: None,
         }
     }
 
+    /// Creates a new, empty context that wraps the previous context.
+    pub fn from_previous_context(context: impl Into<Arc<Context>>) -> Self {
+        Self {
+            type_map: HashMap::new(),
+            previous_context: Some(context.into()),
+        }
+    }
+
+    /// Consumes this `Context` giving back the previous one, if present.
+    pub fn retrieve_previous_context(self) -> Option<Context> {
+        self.previous_context.map(|i| Arc::try_unwrap(i).unwrap())
```

Comment:
The `unwrap` here doesn't seem right. It very well could be that the previous context has a reference count that is higher than 1.  Perhaps this should return an `Option<Arc<Context>>`

Reply:
## TL;DR 

I'll try to get rid of `Arc` since the `Context` in the pipeline does not need to be cloneable.

## Detailed discussion

First of all, I really dislike that we are wrapping a `Context` and returning a `Arc<Context>`. 

Secondly, there is no need, in a pipeline policy, to clone a `Context`. They either read from it or, if they need to mutate it, they first wrap it in a new `Context` and then override that one. At least that's the idea of an immutable `Context` shown here. Given this purpose (like a stack) I would like to be able to "pop" the last `Context` and get back to the previous one.

But I haven't been able to model it properly since `Context` is `clone` so I need to wrap the inner `Context` in an `Arc`. That in turn creates the code you see above (which I agree is wrong 🚶‍♂️🚶‍♂️). 

What I can do is to separate (yet again) the SDK user `Context` (clonable) from a `PipelineContext` (not clonable). `PipelineContext` will own the passed `Context` and possibly the other, overriding `Contexts`. No clone, no arc 🤣.

Thank you for making me think more about this! ❤️